### ISSUE TITLE
Memory Leak from <java class 'java.lang.String'> to python string.

### ISSUE THREAD

**Issue by vwxyzjn** [jpype-project/jpype#856]
Thanks for this amazing project. It allows for much faster data transfer betwen python and Java in our project https://github.com/vwxyzjn/gym-microrts.

However, there seems to be a memory leak problem I am having difficulty debugging with.

In our experiments that run in an AWS machine with 16 GBs of RAM, the following system metrics usually happen over the course of 40 hours:

![image](https://user-images.githubusercontent.com/5555347/95686063-a33dcb00-0bc9-11eb-9477-8feebc8d3941.png)

The system gradually exhausts memory and then CPU utilization goes to zero. Even though the program is still technically running, the program kind of freezes and produce no additional execution apparently. More specifically, our program is scheduled to run 100M timesteps, but after about 55M timesteps, the CPU utilization goes to zero, and the program does not move forward anymore.

I think the python side of the script should not be the problem because it has been used with non-jpype related library and over the course of 300 hours it looks like the following

![image](https://user-images.githubusercontent.com/5555347/95686348-404d3380-0bcb-11eb-81f3-f29cc3f9801e.png)

Therefore I suspect it might be of jpype's issue. To reproduce the issue, try running the following to install gym-microrts

```bash
# make sure you have GIT LFS installed https://git-lfs.github.com/
$ git lfs clone https://github.com/vwxyzjn/gym-microrts.git && \
cd gym-microrts && \
pip install dacite && \
pip install -e .
```

And run the following script. You should see the memory usage gradually increase.

```python
import gym
import gym_microrts
import time
from gym.wrappers import TimeLimit, Monitor
import gc
import tracemalloc
tracemalloc.start()

try:
    env = gym.make("MicrortsDefeatWorkerRushEnemyShaped-v2")
    # env = Monitor(env, f'videos', force=True)
    env.action_space.seed(0)
    env.reset()
    for i in range(100000000):
        action = env.action_space.sample()
        if len((env.unit_location_mask==1).nonzero()[0]) != 0:
            action[0] = (env.unit_location_mask==1).nonzero()[0][0]
        next_obs, reward, done, info = env.step(action)
        if done:
            env.reset()
            current, peak = tracemalloc.get_traced_memory()
            print(f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")
    env.close()
    tracemalloc.stop()
except Exception as e:
    print(e)
    print(e.stacktrace())
```
and you see the results
```
Current memory usage is 1.089761MB; Peak was 1.123628MB
Current memory usage is 1.163414MB; Peak was 1.199529MB
Current memory usage is 1.243207MB; Peak was 1.279058MB
Current memory usage is 1.311169MB; Peak was 1.3459MB
Inconsistent actions were executed!
Worker(45)(1, (2,1), 1, 0) assigned action move(left) at time 220
  Resources: ResourceUsage: 0,0 positions: [11]
Worker(8)(0, (0,1), 1, 0) assigned action move(right) at time 225
  Resources: ResourceUsage: 0,0 positions: [11]
Player resources: 5, 5
Resource Consistency: false
java.lang.Exception: dummy
	at rts.GameState.issue(GameState.java:283)
	at rts.GameState.issueSafe(GameState.java:370)
	at tests.JNIClient.step(JNIClient.java:202)
Current memory usage is 1.387679MB; Peak was 1.424796MB
Current memory usage is 1.459619MB; Peak was 1.496294MB
Current memory usage is 1.534102MB; Peak was 1.569905MB
Current memory usage is 1.604375MB; Peak was 1.640506MB
Current memory usage is 1.677906MB; Peak was 1.713901MB
Current memory usage is 1.755217MB; Peak was 1.787712MB
Current memory usage is 1.821469MB; Peak was 1.858192MB
Inconsistent actions were executed!
Worker(101)(1, (3,1), 1, 0) assigned action move(left) at time 210
  Resources: ResourceUsage: 0,0 positions: [12]
Worker(8)(0, (2,0), 1, 0) assigned action move(down) at time 211
  Resources: ResourceUsage: 0,0 positions: [12]
Player resources: 5, 5
Resource Consistency: false
java.lang.Exception: dummy
	at rts.GameState.issue(GameState.java:283)
	at rts.GameState.issueSafe(GameState.java:370)
	at tests.JNIClient.step(JNIClient.java:202)
Current memory usage is 1.892456MB; Peak was 1.930843MB
Current memory usage is 1.96502MB; Peak was 2.001946MB
Current memory usage is 2.038704MB; Peak was 2.076299MB
Current memory usage is 2.125142MB; Peak was 2.161249MB
Current memory usage is 2.188874MB; Peak was 2.221481MB
Current memory usage is 2.255487MB; Peak was 2.292266MB
Current memory usage is 2.332594MB; Peak was 2.36559MB
Current memory usage is 2.402062MB; Peak was 2.437641MB
Current memory usage is 2.449723MB; Peak was 2.48033MB
Current memory usage is 2.519412MB; Peak was 2.550047MB
```

In comparison, if I run it with a standard environments without jpype
```python
import gym
import gym_microrts
import time
from gym.wrappers import TimeLimit, Monitor
import gc
import tracemalloc
tracemalloc.start()

try:
    env = gym.make("BreakoutNoFrameskip-v4")
    # env = Monitor(env, f'videos', force=True)
    env.action_space.seed(0)
    env.reset()
    for i in range(100000000):
        action = env.action_space.sample()
        next_obs, reward, done, info = env.step(action)
        if done:
            env.reset()
            current, peak = tracemalloc.get_traced_memory()
            print(f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")
    env.close()
    tracemalloc.stop()
except Exception as e:
    print(e)
    print(e.stacktrace())
```
I get a constant memory usuage
```
Current memory usage is 0.658754MB; Peak was 0.872186MB
Current memory usage is 0.667719MB; Peak was 0.872186MB
Current memory usage is 0.667719MB; Peak was 0.872186MB
Current memory usage is 0.668328MB; Peak was 0.872186MB
Current memory usage is 0.668383MB; Peak was 0.872186MB
Current memory usage is 0.66815MB; Peak was 0.872186MB
Current memory usage is 0.668328MB; Peak was 0.872186MB
Current memory usage is 0.668266MB; Peak was 0.872186MB
Current memory usage is 0.668211MB; Peak was 0.872186MB
Current memory usage is 0.668273MB; Peak was 0.872186MB
Current memory usage is 0.668266MB; Peak was 0.872186MB
Current memory usage is 0.668212MB; Peak was 0.872186MB
Current memory usage is 0.668212MB; Peak was 0.872186MB
Current memory usage is 0.668266MB; Peak was 0.872186MB
Current memory usage is 0.668211MB; Peak was 0.872186MB
Current memory usage is 0.668266MB; Peak was 0.872186MB
Current memory usage is 0.668328MB; Peak was 0.872186MB
Current memory usage is 0.668328MB; Peak was 0.872186MB
Current memory usage is 0.66852MB; Peak was 0.872186MB
Current memory usage is 0.668266MB; Peak was 0.872186MB
Current memory usage is 0.668328MB; Peak was 0.872186MB
Current memory usage is 0.668601MB; Peak was 0.872186MB
Current memory usage is 0.668266MB; Peak was 0.872186MB
```

**Comment by vwxyzjn** [comment]
Ok, it appears the culprit is the conversion from `<java class 'java.lang.String'>` to python string. As an example

```python
import jpype
import jpype.imports
from jpype.types import JString
from jpype import JImplements, JOverride, JImplementationFor
jpype.startJVM(convertStrings=False)
import java.lang
import java.util
import tracemalloc
tracemalloc.start()

for i in range(100000000):
    str(JString('{"invalid_action_stats":{"num_invalid_action_null":0,"num_invalid_action_ownership":0,"num_invalid_action_busy_unit":0}}'))
    if i % 2000 ==0:
        current, peak = tracemalloc.get_traced_memory()
        # print(str(response.info))
        print(f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")
tracemalloc.stop()
```

```
Current memory usage is 0.000265MB; Peak was 0.00053MB
Current memory usage is 0.343739MB; Peak was 0.344004MB
Current memory usage is 0.681843MB; Peak was 0.682108MB
Current memory usage is 1.019979MB; Peak was 1.020244MB
Current memory usage is 1.358083MB; Peak was 1.358348MB
Current memory usage is 1.696251MB; Peak was 1.696516MB
Current memory usage is 2.034355MB; Peak was 2.03462MB
Current memory usage is 2.372458MB; Peak was 2.372723MB
Current memory usage is 2.710598MB; Peak was 2.710863MB
Current memory usage is 3.053743MB; Peak was 3.054008MB
Current memory usage is 3.390165MB; Peak was 3.39043MB
Current memory usage is 3.7283MB; Peak was 3.728565MB
Current memory usage is 4.066402MB; Peak was 4.066667MB
Current memory usage is 4.40457MB; Peak was 4.404835MB
Current memory usage is 4.741972MB; Peak was 4.742237MB
Current memory usage is 5.080292MB; Peak was 5.080557MB
Current memory usage is 5.418428MB; Peak was 5.418693MB
Current memory usage is 5.756532MB; Peak was 5.756797MB
Current memory usage is 6.0947MB; Peak was 6.094965MB
Current memory usage is 6.432802MB; Peak was 6.433067MB
Current memory usage is 6.770906MB; Peak was 6.771171MB
Current memory usage is 7.10901MB; Peak was 7.109275MB
Current memory usage is 7.449845MB; Peak was 7.45011MB
Current memory usage is 7.786906MB; Peak was 7.787171MB
Current memory usage is 8.123373MB; Peak was 8.123638MB
Current memory usage is 8.461477MB; Peak was 8.461742MB
Current memory usage is 8.799613MB; Peak was 8.799878MB
Current memory usage is 9.137717MB; Peak was 9.137982MB
Current memory usage is 9.475181MB; Peak was 9.475446MB
Current memory usage is 9.813501MB; Peak was 9.813766MB
Current memory usage is 10.151637MB; Peak was 10.151902MB
Current memory usage is 10.489743MB; Peak was 10.490008MB
Current memory usage is 10.827913MB; Peak was 10.828178MB
Current memory usage is 11.166019MB; Peak was 11.166284MB
Current memory usage is 11.504125MB; Peak was 11.50439MB
Current memory usage is 11.844406MB; Peak was 11.844671MB
Current memory usage is 12.181533MB; Peak was 12.181798MB
```

**Comment by Thrameos** [comment]
Thanks for the report.  I will investigate this week.

**Comment by pelson** [comment]
Great report @vwxyzjn (nice simple example to reproduce)! @Thrameos, I tried running the GC on both the Java and the Python side and see that there is indeed a memory leak. Adapting the code above to include the GC:

```
import jpype
import jpype.imports
from jpype.types import JString
import gc
import tracemalloc


jpype.startJVM(convertStrings=False)
tracemalloc.start()

for i in range(1000000):
    str(JString('A string'))
    if i % 2000 ==0:
        current, peak = tracemalloc.get_traced_memory()
        print(f"Current memory usage is {current / 10**6}MB; Peak was {peak / 10**6}MB")
        jpype.java.lang.System.gc()
        gc.collect()

tracemalloc.stop()
```

Interestingly it is the ``str``  of a ``JString`` which appears to be causing the issue, simply creating a ``JString`` doesn't trigger the leak.

There is a special case for casting JString to a Python string, and it includes some caching. (https://github.com/jpype-project/jpype/blob/v1.0.2/native/python/pyjp_value.cpp#L181-L201). Disabling the caching removes the leak.

Indeed, removing the INCREF on the cache object (https://github.com/jpype-project/jpype/blob/v1.0.2/native/python/pyjp_value.cpp#L198) also fixes the leak. I don't know the Python C-API well enough to know if this makes sense, but the [PyDict_SetItemString](https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItemString) *doesn't* steal the reference, so I think that means we should be decrefing the cache after the call (https://docs.python.org/2.0/api/refcountDetails.html), which is cancelled out by the necessary incref when returning the cache object.

**Comment by pelson** [comment]
@vwxyzjn - this was fixed in #858 and should be part of v1.1 (tracked in #861). If you find any more leaks please report back, but at least the one you managed to isolate is fixed :+1:
